doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+
+- get rid of the NumPy-specific templating language for more expressive C++
+- this would make using sizeless intrinsics (like for SVE) easier.
+
+In addition, we see the following considerations:
+
+- If we use Highway, we would need to switch the C++ wrappers from universal
+  intrinsics to Highway. On the other hand, the work to move to C++ is not
+  complete.
+- If we use Highway, we'd need to rewrite existing kernels using Highway
+  intrinsics. But again, moving to C++ requires touching all those kernels
+  anyway.
+- One concern regarding Highway was whether it is possible to obtain a function
+  pointer for an architecture-specific function instead of calling that
+  function directly. This so that we can be sure that calling 1-D inner loop
+  many times for a single Python API invocation does not incur the dispatching
+  overhead many times. This was investigated: this can be done with Highway
+  too.
+- A second concern was whether it's possible with Highway to allow the user at
+  runtime to select or disable dispatching to certain instruction sets. This is
+  possible.
+- Use of tags in the C++ implementation (Sayed's concern). TODO: describe in
+  more detail.
+
+
+The ``_simd`` unit testing module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rewriting the ``_simd testing`` module to use C++ was done very recently in PR
+`gh-24069`_. It depends on the main PR for the move to C++, `gh-21057`_.
+It allows one to access the C++ intrinsics with almost the same signature, but
+from Python. This is a great way not only for testing, but also for designing
+new SIMD kernels.
+
+It may be possible to add a similar testing and prototyping feature to Highway
+(which uses plain ``googletest``), however currently the NumPy way is quite a
+bit nicer.
+
+
+Math routines
+~~~~~~~~~~~~~
+
+Math or numerical routines are written at a higher level of abstraction than
+the universal intrinsics that are the main focus of this NEP. Highway has only
+a limited number of math routines, and they are not precise enough for NumPy's
+needs. So either way, NumPy's existing routines (which use universal
+intrinsics) will stay, and if we go the Highway route they'll simply have to
+use Highway primitives internally.
+
+There may be other libraries that have numerical routines that can be reused in
+NumPy (e.g., from SLEEF, or perhaps from JPEG XL or some other Highway-using
+libraries). There may be a small benefit here, but likely it doesn't matter too
+much.
+
+
+Supported and missing intrinsics
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some specific intrinsics that NumPy needs may be missing from Highway, e.g.:
+
+- Supporting partial operations on both contiguous and non-contiguous memory
+  access, with both single and pair offsets. Pair offsets is crucial for
+  complex kernels that require 64-bit/32-bit stride alignment. Identified
+  intrinsics related to this:
+
+  - Highway's ``LoadInterleaved2`` would have to be extended with ``*Till``
```

Comment:
```suggestion
Highway needs to implement the following memory operations, which mainly deal with non-contiguous memory operations and partial load/store.

     | Name             |u8 |i8 |u16|i16|u32|i32|u64|i64|f32|f64|
     |:-----------------|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
     | LoadPairTill     | - | - | - | - | x | x | x | x | x | x |
     | LoadTill         | - | - | - | - | x | x | x | x | x | x |
     | IsLoadable       | - | - | - | - | x | x | x | x | x | x |
     | IsStorable       | - | - | - | - | x | x | x | x | x | x |
     | Loadn            | - | - | - | - | x | x | x | x | x | x |
     | LoadnTill        | - | - | - | - | x | x | x | x | x | x |
     | LoadnPair        | - | - | - | - | x | x | x | x | x | x |
     | LoadnPairTill    | - | - | - | - | x | x | x | x | x | x |
     | Storen           | - | - | - | - | x | x | x | x | x | x |
     | StorenTill       | - | - | - | - | x | x | x | x | x | x |
     | StorenPair       | - | - | - | - | x | x | x | x | x | x |
     | StorenPairTill   | - | - | - | - | x | x | x | x | x | x |
     | Lookup128        | - | - | - | - | x | x | x | x | x | x |
```
These operations are described in the new C++ reference section [memory]( https://output.circle-artifacts.com/output/job/6a58c19b-2137-4b00-80a7-50e532c4d7b3/artifacts/0/doc/build/html/reference/simd/intrinsics/memory.html).

Furthermore, Highway needs to identify the behavior of comparisons, including whether they are signaling/non-signaling or ordered and unordered. This information is fully described in the new reference. Please refer to the [comparison](https://output.circle-artifacts.com/output/job/6a58c19b-2137-4b00-80a7-50e532c4d7b3/artifacts/0/doc/build/html/reference/simd/intrinsics/comparison.html) reference for detailed information.

Another aspect that seems to be missing is the integer division by scalar. This operation is fully described [here](https://output.circle-artifacts.com/output/job/6a58c19b-2137-4b00-80a7-50e532c4d7b3/artifacts/0/doc/build/html/reference/simd/intrinsics/arithmetic.html#_CPPv4I00E3Div4TVecRK4TVecRK4Vec3I5TLaneE) as well.

Reply:
Looks like the CI output link no longer works. Is there some other way to see your nicely formatted documentation?

Regarding comparisons, Highway uses _OQ, so ordered and quiet/non-signalling. Would that work for our purposes?

FYI float16 support is [mostly done](https://github.com/google/highway/pull/1608) but not yet tested and requires some more work (casts) for GCC.

We didn't add integer division to Highway because it is generally inefficient and we haven't yet needed it. Is it commonly used in numpy? One option if the division is by a few constant values is to use libdivide, or a compiler, to translate that into vectorized mul by the inverse.

Reply:
But the nep link on the top line of the rendered artifacts point to the absolute NEP path, not the relative one. [here](https://output.circle-artifacts.com/output/job/1e437a73-c8a9-4046-9b4b-3989c816f3ce/artifacts/0/doc/neps/_build/html/nep-0054-simd-cpp-highway.html) is the latest version, that I got from the [CircleCI workflow artifact listing](https://app.circleci.com/pipelines/github/numpy/numpy/20310/workflows/7b238750-f9eb-4b32-8a20-a829db3d2f54/jobs/32828/artifacts)

Reply:
I would prefer a high-level summary of missing functionality. How about:

- Highway is missing non-contiguous memory operations and partial load/store operations.
- Highway needs to identify the behavior of comparisons, including whether they are signaling/non-signaling or ordered and unordered.
- Highway is missing integer division by a scalar.

Reply:
The latest version of PR #2107 's documentation is [here](https://output.circle-artifacts.com/output/job/8ba4034e-8602-48ae-8678-f7e839dd9edf/artifacts/0/doc/build/html/reference/simd/intrinsics/index.html)

Reply:
Updates:
- we have non-contiguous ops, but they use Gather/Scatter and are thus slow. What's missing is special-cases such as pairs of doubles using 128-bit loads.
- as of last week we have partial Load/Store and also LoadNOr.

Comparisons, reductions and integer div are indeed missing.

Reply:
> I would prefer a high-level summary of missing functionality. How about:

Here's a list of missing intrinsics:

- Vec<TLane> LoadPairTill(const TLane *ptr, size_t plen, TLane fill0, TLane fill1)
    Partially pair loads N-lane vector contents from memory.

- Vec<TLane> Loadn(const TLane *ptr, intptr_t stride)
    Load N-lane vector contents from non-contiguous memory block.

- Vec<TLane> LoadnTill(const TLane *ptr, intptr_t stride, size_t len, TLane fill)
    Partially loads N-lane vector contents from a non-contiguous memory block.

- Vec<TLane> LoadnPair(const TLane *ptr, intptr_t stride)
    Pair loads N-lane vector contents from non-contiguous memory block.

- Vec<TLane> LoadnPairTill(const TLane *ptr, intptr_t stride, size_t plen, TLane fill0, TLane fill1)
    Partially pair loads N-lane vector contents from non-contiguous memory block.

- void StorePairTill(TLane *ptr, size_t plen, const Vec<TLane> &a)
    Partially pair store N-lane vector contents to memory.

- void Storen(TLane *ptr, intptr_t stride, const Vec<TLane> &a)
    Store N-lane vector contents to non-contiguous memory.

- void StorenTill(TLane *ptr, intptr_t stride, size_t len, const Vec<TLane> &a)
    Partially store N-lane vector contents to non-contiguous memory.

- void StorenPair(TLane *ptr, intptr_t stride, const Vec<TLane> &a)
    Pair store N-lane vector contents to non-contiguous memory.

- void StorenPairTill(TLane *ptr, intptr_t stride, size_t plen, const Vec<TLane> &a)
    Partially pair store N-lane vector contents to non-contiguous memory.

- Vec<TLane> Lookup128(const TLane *table, const Vec<MakeUnsigned<TLane>> &idx)
    Perform a lookup operation on a 128-byte table using lane indices.

- TVec MaxProp(const TVec &a, const TVec &b)
    Maximum with Floating-Point with Non-NaN Propagation.

- TVec MinProp(const TVec &a, const TVec &b)
    Minimum with Floating-Point with Non-NaN Propagation.

- TVec MaxPropNan(const TVec &a, const TVec &b)
    Maximum with Floating-Point with NaN Propagation.

- TVec MinPropNan(const TVec &a, const TVec &b)
    Minimum with with NaN Propagation.

- TLane ReduceMax(const TVec &a)
    Native Reduce Maximum.

- TLane ReduceMin(const TVec &a)
    Native Reduce Minimum.

- TLane ReduceMaxProp(const TVec &a)
    Reduce Maximum Floating-Point with Non-NaN Propagation.

- TLane ReduceMinProp(const TVec &a)
    Reduce Minimum Floating-Point with Non-NaN Propagation.

- TLane ReduceMaxPropNan(const TVec &a)
    Reduce Maximum Floating-Point with NaN Propagation.

- TLane ReduceMinPropNan(const TVec &a)
    Reduce Minimum Floating-Point with NaN Propagation.

- TLane ReduceMaxProp(const TVec &a)
    Reduce Maximum Floating-Point with Non-NaN Propagation.

- TLane ReduceMinProp(const TVec &a)
    Reduce Minimum Floating-Point with Non-NaN Propagation.

- TLane ReduceMaxPropNan(const TVec &a)
    Reduce Maximum Floating-Point with NaN Propagation.

- TLane ReduceMinPropNan(const TVec &a)
    Reduce Minimum Floating-Point with NaN Propagation.

- Vec3<TLane> Divisor(TLane d)
    Generate the precomputed parameters of the divisor.

- TVec Div(const TVec &a, const Vec3<TLane> &b)
    Performs element-wise integer division between the lanes of the input vector ‘a’ and the scalar value b.

- TVec MulAddSub(const TVec &a, const TVec &b, const TVec &c)
    Multiply the lanes of two vectors, add the lanes of a third vector for odd lanes, and subtract for even lanes.

- TVec IfAdd/IfSub/IfDiv(const TMask &m, const TVec &a, const TVec &b, const TVec &c)
    Conditional add/sub/div of vectors based on a mask.

In addition(not supported yet), I was working in a patch to support the following AVX512 operations which is required in order to replace SVML kernels:
- _mm512_getmant_ps/_pd
- _mm512_scalef_ps/_pd
- _mm512_getexp_ps/_pd


> as of last week we have partial Load/Store and also LoadNOr.

Current Highway partial Load/Store only supports contiguous memory operations according to the following testing unite:
LoadN:
https://github.com/google/highway/blob/ba0900a4957b929390ab73827235557959234fea/hwy/tests/memory_test.cc#L350-L404
StoreN:
https://github.com/google/highway/blob/ba0900a4957b929390ab73827235557959234fea/hwy/tests/memory_test.cc#L572-L634

Reply:
Thank you @seiko2plus for making the list :) I am back in the office and adding it to our list of pending ops.

FYI `ReduceMax/ReduceMin` can already be expressed as `GetLane(Max/MinOfLanes())`.

> Current Highway partial Load/Store only supports contiguous memory operations

We also have MaskedGather/Scatter. The mask can be obtained for partial N via FirstN, right?

Reply:
>We also have MaskedGather/Scatter. The mask can be obtained for partial N via FirstN, right?

if performance regression isn't matter then yes, sure.

Reply:
I think it is too much of a moving target to publish this table in a permanent document. Maybe it belongs in a Highway issue instead, and then we can refer to the issue?